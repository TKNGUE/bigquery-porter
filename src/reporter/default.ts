import process from 'node:process';
import pc from 'picocolors';
import logUpdate from 'log-update';

import { F_CHECK, F_CROSS } from '../../src/figures.js';
import { spyConsole } from '../../src/runtime/console.js';

import type {
  Reporter, ReporterTask,
} from '../../src/types.js';

const spinnerFrames = process.platform === 'win32'
  ? ['-', '\\', '|', '/']
  : ['⠋', '⠙', '⠹', '⠸', '⠼', '⠴', '⠦', '⠧', '⠇', '⠏'];

function elegantSpinner() {
  let index = 0;
  return () => {
    index = ++index % spinnerFrames.length;
    return spinnerFrames[index] ?? '';
  };
}

class DefaultReporter implements Reporter {
  tasks: ReporterTask[] = [];
  hooks: (() => void)[] = [];
  separator: string = '/';
  spinnerMap: WeakMap<ReporterTask, () => string> = new WeakMap();

  report_task(task: ReporterTask): string {
    let s = '';
    let c = pc.red;
    switch (task.status) {
      case 'success':
        s = F_CHECK;
        c = pc.green;
        break;

      case 'failed':
        s = `${F_CROSS}`;
        c = pc.red;
        break;

      case 'running':
        if (!this.spinnerMap.has(task)) {
          this.spinnerMap.set(task, elegantSpinner());
        }
        const spin = this.spinnerMap.get(task);
        if (!spin) {
          throw new Error('Unreachable codes');
        }

        s = spin();
        c = pc.gray;
        break;

      case 'pending':
        return '';
    }

    const title = c(`${s} ${task.name.split(this.separator).pop()}`);
    if (task.error) {
      return `${title}: ${pc.bold(task.error)}`.trim();
    } else {
      const msg = task.message ? ` (${task.message ?? ''})` : '';
      return `${title} ${msg}`.trim();
    }
  }

  report_tree(tasks: ReporterTask[], level = 0, max_level = 4): string {
    const groups = tasks.reduce((acc, t) => {
      const parts = t.name.split(this.separator);
      const key = parts.length === level + 1
        ? '#tail'
        : (parts[level] ?? '#none');
      if (acc.get(key) === undefined) {
        acc.set(key, []);
      }
      acc.get(key)?.push(t);
      return acc;
    }, new Map<string, ReporterTask[]>());

    let s = '';
    const childStingifier = (tasks: ReporterTask[], level: number) => {
      const spaces = '  '.repeat(level);
      const body = tasks.map((t) => this.report_task(t)).filter((s) => s).join(
        '\n' + spaces,
      );

      return `${spaces}${body}\n`;
    };
    for (const [group_key, tasks] of groups) {
      if (group_key == '#tail') {
        s += childStingifier(tasks, level);
      } else {
        s += '  '.repeat(level) + pc.underline(group_key) + '\n';
        if (level < max_level) {
          s += this.report_tree(tasks, level + 1, max_level);
        } else {
          s += childStingifier(tasks, level + 1);
        }
      }
    }
    return s;
  }

  onInit(tasks: ReporterTask[]) {
    this.tasks = tasks;
    this.spinnerMap = new WeakMap();
    logUpdate.done();
    const { buffer, teardown } = spyConsole(() => 'unknown');
    this.hooks.push(
      teardown,
      () => {
        const stdout = buffer.stdout.get('unknown');
        const stderr = buffer.stderr.get('unknown');
        if (stdout && stdout.length > 0) {
          console.log('STDOUT: ');
          for (const data of stdout) {
            console.log(data.toString());
          }
        }

        if (stderr && stderr.length > 0) {
          console.log('STDERR: ');
          for (const data of stderr) {
            console.log(data.toString());
          }
        }
      },
    );
  }

  async onUpdate() {
    const report = this.report_tree(this.tasks);
    const taskCount = this.tasks.filter((t) => !t.done()).length;
    logUpdate(
      `Tasks: remaing ${taskCount}\n` +
      report,
    );
  }

  onFinished() {
    this.hooks.forEach((h) => h());
  }
}

export { DefaultReporter };
