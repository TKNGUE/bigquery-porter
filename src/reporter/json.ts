import type { Reporter, ReporterTask } from '../types.js';

class JSONReporter implements Reporter {
  tasks: ReporterTask[] = [];

  onInit(tasks: ReporterTask[]) {
    this.tasks = tasks;
  }

  async onUpdate() {
  }

  onFinished() {
    for (const task of this.tasks) {
      console.log(JSON.stringify(
        {
          name: task.name,
          status: task.status,
          result: {
            message: task.message,
            error: task.error,
          },
        },
        null,
      ));
    }
  }
}

export { JSONReporter };
