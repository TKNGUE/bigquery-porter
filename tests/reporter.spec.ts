import { afterEach, describe, expect, it, vi } from 'vitest';
import { DefaultReporter } from '../src/reporter/default.js';
import { JSONReporter } from '../src/reporter/json.js';
import { Task } from '../src/tasks/base.js';

describe('Reporter: Default Reporter', () => {
  it.concurrent('Should test', async () => {
    const reporter = new DefaultReporter();
    const tasks = [
      new Task(
        'test1',
        async () => {
          await new Promise((resolve) => setTimeout(resolve, 10));
          return 'test';
        },
      ),
    ];
    tasks.forEach((t) => t.run());

    reporter.onInit(tasks);
    while (tasks.some((t) => !t.done())) {
      await new Promise((resolve) => setTimeout(resolve, 20));
    }
    reporter.onUpdate();
    reporter.onFinished();
  });

  it.concurrent('Too long message tasks', async () => {
    const reporter = new DefaultReporter();
    const tasks = [
      new Task(
        'success'.repeat(100),
        async () => {
          await new Promise((resolve) => setTimeout(resolve, 10));
          return 'msg';
        },
      ),
      new Task(
        'failed',
        async () => {
          throw new Error('msg '.repeat(100));
        },
      ),
    ];
    tasks.forEach((t) => t.run());

    reporter.onInit(tasks);
    while (tasks.some((t) => !t.done())) {
      reporter.onUpdate();
      await new Promise((resolve) => setTimeout(resolve, 20));
    }
    reporter.onUpdate();
    reporter.onFinished();
  });
});

describe('Reporter: JSON Reporter', () => {
  afterEach(() => {
    vi.resetAllMocks();
  });

  it('core', async () => {
    const out = [];
    vi.spyOn(console, 'log').mockImplementation((s) => {
      out.push(s);
    });

    const reporter = new JSONReporter();
    const tasks = [
      new Task(
        'success task',
        () => new Promise((resolve) => resolve('1')),
      ),
      new Task(
        'failed task',
        () => new Promise((_, reject) => reject(new Error('error'))),
      ),
    ];
    tasks.forEach((t) => t.run());

    reporter.onInit(tasks);
    while (tasks.some((t) => !t.done())) {
      await reporter.onUpdate();
    }
    reporter.onFinished();

    expect(out).toMatchSnapshot();
  });
});
