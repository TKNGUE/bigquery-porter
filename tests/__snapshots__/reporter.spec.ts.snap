// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Reporter: JSON Reporter > core 1`] = `
[
  "{\\"name\\":\\"success task\\",\\"result\\":\\"1\\"}",
  "{\\"name\\":\\"failed task\\",\\"error\\":\\"error\\"}",
]
`;
